### ISSUE TITLE
question about shuffling the data before splitting them into train and test dataloader

### ISSUE THREAD

**Issue by turmeric-blend** [jankrepl/deepdow#86]
Hi, I have a question about shuffling the data before splitting them into train and test dataloader.

Currently, the flow of data preparation would be:
1. get `X` and `y` from `raw_to_Xy` where the data would still be ordered in time (index 0 = oldest date, index -1 = newest date)
2. getting the index array from `X`
3. splitting the index array (eg first 80% to train, last 20% to test)
4. getting dataset from `InRAMDataset`
5. getting train/test dataloader from the respective index

For this case, `dataloader_train` would still have the first 80% of dates in (eg 2011 to 2018) and `dataloader_test` having the last 20% of dates (eg 2018 to 2020)

I was thinking since the inputs to the network are **returns**, and hence **returns** tend to be in the same distribution no matter the period (at least for _daily_ returns), why not shuffle the data before splitting it into the dataloader, similar to how images are prepared for computer vision tasks, or any task that are time-independent in general (of course there is still a time-dependent component per sample which is not affected by this process). 

So the dataloaders would look something like (`dataloader_train` having dates 2011, 2013, 2014, 2015, 2016, 2017, 2018, 2020) and (`dataloader_test` having dates 2012, 2019) ..etc.

This way, the learned model would not be so bias to the test data dates (which are cramped to be only the more present dates) and be more general. Does this sound right? Would be good to have your opinion on it.

Of course, this would be just an option as inputs may not be in returns.

**Comment by jankrepl** [comment]
One important thing to note is that both the `RigidDataLoader` and `FlexibleDataLoader` automatically shuffle the the samples. See the below link
https://github.com/jankrepl/deepdow/blob/f123a8212ab3b6a2ff8514dfd8b936bc47a8806d/deepdow/data/load.py#L303

That means that even if your indices are consecutive numbers (e.g. representing consecutive days) the actual generated batches will mosts likely have nonconsecutive samples. So to come back to what you wrote, the assumption of independence within the same DataLoader is already embedded in `deepdow`.

Regarding creation of the train test split, I would say that you need to make that decision yourself. I guess in the documentation and examples I just wanted to make sure that train and test set come from totally different time periods. There are 2 reasons:
* Without reflecting too much on statistical properties of the underlying time series, ideally I would like a model that was trained on historical data to do well on future data. Because that is what really matters. That is why it makes sense to create a train test split that mirrors this requirement.
* Shuffling randomly before the split might actually leak a lot of information from the training set to the test set. If your `lookback=10` for example, then it will take 10 time periods for the feature tensor `X` to contain no information that happened on a given date. All this because of the rolling window approach. It is analogous to taking an image (of shape `(lookback, n_assets)`) and shifting it by one pixel each time period and putting some of augmented images in the training and the remaining ones in the test set. Technically, the training and test set do not contain exact pixel-by-pixel duplicates, however, networks like CNN do not really care about these shifts and one is leaking information between the two sets.

**Comment by turmeric-blend** [comment]
yes what you said makes sense as well, thanks for your patience!
